<launch>
  <!-- Arguments -->
  <arg name="model" default="waffle_pi"/>
  <arg name="bag_name" default="fixed_slam_easy"/>
  <arg name="map_file" default="$(find turtlebot3_datasets)/data/map.yaml" />

  <!-- Turtlebot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <!-- Map server -->
  <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)" />

  <!-- TF -->
  <node pkg="tf2_ros" type="static_transform_publisher" name="initial_tf" args="0.935 1.34 -0.023 0.001 -0.003 0.737 0.676 mocap map"/>
  <node pkg="tf2_ros" type="static_transform_publisher" name="mocap_to_odom" required="true" args="0 0 0 0 0 0 map odom"/>
  
  <!-- ROSBAG -->
  <param name="/use_sim_time" value="true"/>
  <node pkg="rosbag" type="play" name="rosbag_player" 
          args="--clock $(find turtlebot3_datasets)/data/$(arg bag_name).bag"/>

    <node pkg="robot_localization" type="ekf_localization_node" name="robot_localization_ekf_node_odom" clear_params="true">
  <param name="frequency" value="10" />
  <param name="sensor_timeout" value="0.2" />
  <param name="two_d_mode" value="true" />
  <param name="publish_tf" value="false" />
  <param name="map_frame" value="map" />
  <param name="odom_frame" value="odom" />
  <param name="base_link_frame" value="base_footprint" />
  <param name="world_frame" value="map" />

  <!-- Odometry -->
  <param name="odom0" value="odom" />
  <rosparam param="odom0_config">
    [true, true, false,
     false, false, true,
     true, true, false,
     false, false, true,
     false, false, false]
  </rosparam>
  <param name="odom0_relative" value="false"/>

  <!-- IMU -->
  <param name="imu0" value="imu" />
    <rosparam param="imu0_config">
      [false, false, false,
       false, false, true,
       false, false, false,
       true, true, false,
       false, false, false]</rosparam>
    <param name="imu0_relative" value="true"/>

  <!-- Ground Truth (PoseWithCovariance from gtInfo) -->
  <param name="gt0" value="/gtInfo" />
  <rosparam param="gt0_config">
    [true, true, false,  # X, Y, Z positions
     false, false, true, # Roll, Pitch, Yaw orientations
     false, false, false, # X, Y, Z velocities
     false, false, false, # Roll, Pitch, Yaw angular velocities
     false, false, false] # Linear acceleration (if available)
  </rosparam>
  <param name="gt0_differential" value="false" />
</node>
  
  <!-- Ground Truth Trajectory -->

  <!-- AMCL-->

  <!-- Groundtruth path -->
  <node name="ground_truth_publisher" pkg="turtlebot3_datasets" type="publish_pose.py">
  </node>
  <!-- Estimated path -->
  <node name="path_publisher" pkg="turtlebot3_datasets" type="publish_path_odom.py">
  </node>

  <!-- Position error publisher -->

  <!-- rviz -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find turtlebot3_datasets)/data/turtlebot3_visualize_localization_sim.rviz"/>

</launch>
